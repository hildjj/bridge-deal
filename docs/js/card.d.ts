export declare enum Suit {
    CLUBS = "\u2663",
    DIAMONDS = "\u2662",
    HEARTS = "\u2661",
    SPADES = "\u2660"
}
export type SuitStr = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export declare const RankValues: {
    2: number;
    3: number;
    4: number;
    5: number;
    6: number;
    7: number;
    8: number;
    9: number;
    T: number;
    J: number;
    Q: number;
    K: number;
    A: number;
};
export type Rank = keyof typeof RankValues;
export interface Shape {
    spades: number;
    hearts: number;
    diamonds: number;
    clubs: number;
}
export declare class Inspected {
}
export declare class Card extends Inspected {
    static POINTS: {
        [key: string]: number;
    };
    rank: Rank;
    suit: Suit;
    index: number;
    points: number;
    constructor(rank: Rank, suit: Suit, index: number);
    toString(): string;
}
export declare class Hand extends Inspected {
    cards: Card[];
    name: string;
    constructor(name: string);
    get points(): number;
    get spades(): Card[];
    get hearts(): Card[];
    get diamonds(): Card[];
    get clubs(): Card[];
    get shape(): Shape;
    get needed(): bigint;
    push(cd: Card): void;
    lin(): string;
    toString(): string;
}
export declare class Deal extends Inspected {
    private static D;
    num: bigint;
    hands: Hand[];
    constructor(num?: bigint);
    get north(): Hand;
    get east(): Hand;
    get south(): Hand;
    get west(): Hand;
    static weight(cards: Card[], weights: number[]): number;
    lin(): string;
    toString(): string;
}
export type DealPredicate = (deal: Deal) => boolean;
export declare function findDeal(filter?: DealPredicate): [Deal, number];
export declare function deals(num: number, filter?: DealPredicate): Deal[];
export declare function prec2d(d: Deal): boolean;
