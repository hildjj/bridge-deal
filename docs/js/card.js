const MAX_TRIES = 1000000;
export var Suit;
(function (Suit) {
    Suit["CLUBS"] = "\u2663";
    Suit["DIAMONDS"] = "\u2662";
    Suit["HEARTS"] = "\u2661";
    Suit["SPADES"] = "\u2660";
})(Suit || (Suit = {}));
export var NoTrump;
(function (NoTrump) {
    NoTrump["NT"] = "N";
})(NoTrump || (NoTrump = {}));
export const BidSuit = { ...Suit, ...NoTrump };
export var Vuln;
(function (Vuln) {
    Vuln["ALL"] = "All";
    Vuln["NONE"] = "None";
    Vuln["NS"] = "NS";
    Vuln["EW"] = "EW";
})(Vuln || (Vuln = {}));
export var Direction;
(function (Direction) {
    Direction["NORTH"] = "north";
    Direction["EAST"] = "east";
    Direction["SOUTH"] = "south";
    Direction["WEST"] = "west";
})(Direction || (Direction = {}));
let rnk = 0;
export const RankValues = {
    2: rnk++,
    3: rnk++,
    4: rnk++,
    5: rnk++,
    6: rnk++,
    7: rnk++,
    8: rnk++,
    9: rnk++,
    T: rnk++,
    J: rnk++,
    Q: rnk++,
    K: rnk++,
    A: rnk++,
};
export class Inspected {
    [Symbol.for('Deno.customInspect')]() {
        return this.toString();
    }
    [Symbol.for('nodejs.util.inspect.custom')]() {
        return this.toString();
    }
}
export class Ref extends Inspected {
    ref;
    constructor(ref) {
        super();
        this.ref = ref;
    }
    toJSON() {
        return { ref: this.ref };
    }
    toString() {
        return `\${${this.ref}}`;
    }
}
export class Card extends Inspected {
    static POINTS = { A: 4, K: 3, Q: 2, J: 1 };
    rank;
    suit;
    index;
    points;
    constructor(rank, suit, index) {
        super();
        this.rank = rank;
        this.suit = suit;
        this.index = index;
        this.points = Card.POINTS[this.rank] ?? 0;
    }
    toString() {
        return `${this.rank}${this.suit}`;
    }
}
rnk = 51;
export const Deck = Object
    .values(Suit)
    .flatMap(s => Object.keys(RankValues)
    .map(r => new Card(r, s, rnk--)));
export class Hand extends Inspected {
    cards = [];
    name;
    dir;
    #points = undefined;
    #suits = undefined;
    constructor(name) {
        super();
        this.name = name;
        this.dir = name.toLowerCase();
    }
    get points() {
        if (this.#points === undefined) {
            this.#points = this.cards.reduce((t, v) => t + v.points, 0);
        }
        return this.#points;
    }
    get spades() {
        return this.#getSuits()[Suit.SPADES];
    }
    get hearts() {
        return this.#getSuits()[Suit.HEARTS];
    }
    get diamonds() {
        return this.#getSuits()[Suit.DIAMONDS];
    }
    get clubs() {
        return this.#getSuits()[Suit.CLUBS];
    }
    get shape() {
        const s = this.#getSuits();
        return {
            spades: s[Suit.SPADES].length,
            hearts: s[Suit.HEARTS].length,
            diamonds: s[Suit.DIAMONDS].length,
            clubs: s[Suit.CLUBS].length,
        };
    }
    get shapeAny() {
        return Object
            .values(this.shape)
            .sort((a, b) => b - a);
    }
    get needed() {
        return 13n - BigInt(this.cards.length);
    }
    static ranks(cards) {
        return cards.map(s => s.rank).join('');
    }
    shapeStr() {
        return Object
            .values(this.shape)
            .join('');
    }
    isShape(s, h, d, c) {
        const { shape } = this;
        return (shape.spades === s) &&
            (shape.hearts === h) &&
            (shape.diamonds === d) &&
            (shape.clubs === c);
    }
    isShapeAny(...nums) {
        const sa = this.shapeAny;
        const nums_len = nums.length;
        for (let i = 0; i < nums_len; i++) {
            if (sa[i] !== nums[i]) {
                return false;
            }
        }
        return true;
    }
    hasVoid() {
        const sa = this.shapeAny;
        return sa[3] === 0;
    }
    hasSingleton() {
        const sa = this.shapeAny;
        return sa.includes(1);
    }
    hasSingletonOrVoid() {
        const sa = this.shapeAny;
        return sa.findIndex(s => (s === 0) || (s === 1)) !== -1;
    }
    hasCards(cards) {
        const suit = {
            C: this.clubs,
            D: this.diamonds,
            H: this.hearts,
            S: this.spades,
        }[cards[0]];
        for (const c of cards.slice(1)) {
            if (!suit?.some(card => card.rank === c)) {
                return false;
            }
        }
        return true;
    }
    *suits() {
        yield ['Spades', this.spades];
        yield ['Hearts', this.hearts];
        yield ['Diamonds', this.diamonds];
        yield ['Clubs', this.clubs];
    }
    ranks() {
        return [
            Hand.ranks(this.spades),
            Hand.ranks(this.hearts),
            Hand.ranks(this.diamonds),
            Hand.ranks(this.clubs),
        ];
    }
    range(min, max) {
        const pts = this.points;
        return (pts >= min) && (pts <= max);
    }
    balanced() {
        const { shapeAny } = this;
        return (shapeAny[0] < 6) && (shapeAny[3] > 1) &&
            !((shapeAny[0] === 5) && (shapeAny[1] === 4));
    }
    balancedNoM() {
        return this.balanced() && this.spades.length < 5 && this.hearts.length < 5;
    }
    push(cd) {
        this.cards.push(cd);
    }
    pbn() {
        return this.ranks().join('.');
    }
    lin() {
        return `S${Hand.ranks(this.spades)}H${Hand.ranks(this.hearts)}D${Hand.ranks(this.diamonds)}C${Hand.ranks(this.clubs)}`;
    }
    toString() {
        return `${Object
            .values(Suit)
            .map(s => `${s}: ${this.cards
            .filter(cd => cd.suit === s)
            .map(cd => cd.rank)
            .join('')}`)
            .join(' ')} (${this.points})`;
    }
    #getSuits() {
        if (!this.#suits) {
            this.#suits = {
                [Suit.SPADES]: [],
                [Suit.HEARTS]: [],
                [Suit.DIAMONDS]: [],
                [Suit.CLUBS]: [],
            };
            for (const c of this.cards) {
                this.#suits[c.suit].push(c);
            }
        }
        return this.#suits;
    }
}
export class Bid extends Inspected {
    static PASS = 0;
    static DOUBLE = -1;
    static REDOUBLE = -2;
    static BidFromName = {
        p: Bid.PASS,
        P: Bid.PASS,
        X: Bid.DOUBLE,
        x: Bid.DOUBLE,
        XX: Bid.REDOUBLE,
        xx: Bid.REDOUBLE,
    };
    level;
    suit;
    alert;
    description;
    constructor(opts) {
        super();
        if (!opts) {
            opts = { level: Bid.PASS };
        }
        if (typeof opts !== 'object') {
            throw new TypeError(`Invalid opts: "${opts}"`);
        }
        if (typeof opts.level === 'string') {
            const lev = parseInt(opts.level, 10);
            if (isNaN(lev)) {
                throw new Error(`Invalid level: "${opts.level}"`);
            }
            opts.level = lev;
        }
        if (typeof opts.level === 'number') {
            if (opts.suit) {
                if ((opts.level < 1) || (opts.level > 7)) {
                    throw new Error(`Invalid level: "${opts.level}"`);
                }
            }
            else if ((opts.level > Bid.PASS) || (opts.level < Bid.REDOUBLE)) {
                throw new Error(`Invalid level: "${opts.level}"`);
            }
        }
        this.level = opts.level;
        if (typeof opts.suit === 'string') {
            this.suit = {
                C: BidSuit.CLUBS,
                D: BidSuit.DIAMONDS,
                H: BidSuit.HEARTS,
                S: BidSuit.SPADES,
                N: BidSuit.NT,
                [BidSuit.CLUBS]: BidSuit.CLUBS,
                [BidSuit.DIAMONDS]: BidSuit.DIAMONDS,
                [BidSuit.HEARTS]: BidSuit.HEARTS,
                [BidSuit.SPADES]: BidSuit.SPADES,
            }[opts.suit.toUpperCase()];
            if (!this.suit) {
                throw new Error(`Invalid suit lookup: "${opts.suit.toUpperCase()}"`);
            }
        }
        else {
            this.suit = opts.suit;
        }
        this.alert = opts.alert ?? false;
        this.description = opts.description;
    }
    toString(bare = false) {
        let ret = '';
        switch (this.level) {
            case -2:
                ret = 'XX';
                break;
            case -1:
                ret = 'X';
                break;
            case 0:
                ret = 'P';
                break;
            default:
                ret = `${this.level}${this.suit}`;
                break;
        }
        if (!bare) {
            if (this.alert) {
                ret += '!';
            }
            if (this.description) {
                ret += ': ';
                ret += this.description;
            }
        }
        return ret;
    }
    toJSON() {
        return {
            level: this.level,
            suit: this.suit,
            alert: this.alert,
            description: this.description,
        };
    }
    serialize() {
        let ret = '{';
        if (this.level instanceof Ref) {
            if (this.level.ref === 'level') {
                ret += 'level,';
            }
            else {
                ret += `level: ${this.level.ref},`;
            }
        }
        else {
            ret += `level: ${this.level},`;
        }
        if (this.suit) {
            if (this.suit instanceof Ref) {
                if (this.suit.ref === 'suit') {
                    ret += 'suit,';
                }
                else {
                    ret += `suit: ${this.suit.ref},`;
                }
            }
            else {
                ret += `suit: '${this.suit}',`;
            }
        }
        if (this.alert) {
            ret += 'alert: true,';
        }
        if (this.description) {
            ret += `description: \`${this.description}\`,`;
        }
        ret += '}';
        return ret;
    }
}
export class Deal extends Inspected {
    static D = 53644737765488792839237440000n;
    static randBuf = new Uint8Array(16);
    static randBI = new BigUint64Array(this.randBuf.buffer, this.randBuf.byteOffset, this.randBuf.byteLength / BigUint64Array.BYTES_PER_ELEMENT);
    num;
    vuln = Vuln.NONE;
    dealer = Direction.NORTH;
    bids = [];
    hands = [
        new Hand('North'), new Hand('East'), new Hand('South'), new Hand('West'),
    ];
    constructor(num) {
        super();
        if ((num === undefined) || (num < 0n)) {
            num = Deal.randD();
        }
        this.num = num;
        let K = Deal.D;
        let I = this.num;
        for (let C = 52n; C > 0; C--) {
            let X = 0n;
            for (const h of this.hands) {
                I -= X;
                X = K * h.needed / C;
                if (I < X) {
                    h.push(Deck[Number(C) - 1]);
                    break;
                }
            }
            K = X;
        }
    }
    get north() {
        return this.hands[0];
    }
    get east() {
        return this.hands[1];
    }
    get south() {
        return this.hands[2];
    }
    get west() {
        return this.hands[3];
    }
    static fromJSON(o) {
        if (typeof o === 'string') {
            o = JSON.parse(o);
        }
        const d = new Deal(BigInt(`0x${o.num}`));
        o.names.forEach((v, i) => {
            d.hands[i].name = v;
        });
        d.vuln = o.vuln;
        d.dealer = o.dealer;
        d.bids = o.bids.map(b => new Bid(b));
        return d;
    }
    static weight(cards, weights) {
        let tot = 0;
        for (const cd of cards) {
            const pos = 12 - RankValues[cd.rank];
            if (pos >= weights.length) {
                break;
            }
            tot += weights[pos];
        }
        return tot;
    }
    static randD() {
        crypto.getRandomValues(this.randBuf);
        return ((this.randBI[0] << 64n) | this.randBI[1]) % this.D;
    }
    randVuln() {
        const v = Math.floor(Math.random() * 4);
        this.vuln = Object.values(Vuln)[v];
    }
    isVulnerable(hand) {
        switch (this.vuln) {
            case Vuln.ALL:
                return true;
            case Vuln.NONE:
                return false;
            case Vuln.NS:
                return (hand.dir === Direction.NORTH) || (hand.dir === Direction.SOUTH);
            case Vuln.EW:
                return (hand.dir === Direction.EAST) || (hand.dir === Direction.WEST);
        }
        throw new TypeError(`Invalid vulnerability: "${this.vuln}"`);
    }
    bid(opts) {
        this.bids.push(new Bid(opts));
    }
    lin() {
        return `https://www.bridgebase.com/tools/handviewer.html?lin=qx|o${this.num}|md|3${this.south.lin()},${this.west.lin()},${this.north.lin()}|rh||ah|${this.num}|sv|0|pg||`;
    }
    pbn() {
        return `N:${this.north.pbn()} ${this.east.pbn()} ${this.south.pbn()} ${this.west.pbn()}`;
    }
    toString() {
        let ret = `Dealer: ${this.dealer}, Vuln: ${this.vuln}\n`;
        ret += this.hands
            .map(h => `${h.name}: ${h.toString()}`)
            .join('\n');
        ret += '\n';
        if (this.bids.length) {
            ret += this.bids.map(b => b.toString()).join('-');
            ret += '\n';
        }
        return ret;
    }
    toJSON() {
        return {
            num: this.num.toString(16),
            vuln: this.vuln,
            dealer: this.dealer,
            bids: this.bids.map(b => b.toJSON()),
            names: this.hands.map(h => h.name),
        };
    }
}
export function findDeal(filter, maxTries = MAX_TRIES) {
    let tries = 0;
    while (true) {
        if (tries++ > maxTries) {
            throw new Error(`Too many tries: ${tries}`);
        }
        const d = new Deal();
        if (!filter) {
            return [d, tries];
        }
        const ret = filter.call({}, d, Deal);
        if (typeof ret !== 'boolean') {
            throw new Error(`Invalid return: "${ret}"`);
        }
        if (ret) {
            return [d, tries];
        }
    }
}
export function deals(num, filter) {
    const ret = new Array(num);
    let count = 0;
    while (count < num) {
        [ret[count++]] = findDeal(filter);
    }
    return ret;
}
