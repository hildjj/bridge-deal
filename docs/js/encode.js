async function bytesToBase64url(blob) {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onload = (event) => {
            const dataUrl = reader.result;
            if (typeof dataUrl === 'string') {
                const [_, base64] = dataUrl.split(',');
                resolve(base64.replace(/[+/]/g, (s) => ({
                    '+': '-',
                    '/': '_',
                    '=': '',
                }[s] ?? s)));
            }
            else {
                reject(new Error('Invalid FileReader onload event'));
            }
        };
        reader.readAsDataURL(blob);
    });
}
export async function compressString(txt) {
    const te = new TextEncoderStream();
    const rd = te.readable.pipeThrough(new CompressionStream('deflate'));
    const w = te.writable.getWriter();
    await w.write(txt);
    await w.close();
    const r = new Response(rd);
    return bytesToBase64url(await r.blob());
}
export async function decompressString(compressed) {
    const b64 = compressed.replace(/[_-]/g, (s) => ({
        '-': '+',
        '_': '/',
    }[s] ?? s));
    const resp = await fetch(`data:text/javascript;base64,${b64}`);
    const dec = resp.body?.pipeThrough(new DecompressionStream('deflate'));
    const res = new Response(dec);
    return res.text();
}
