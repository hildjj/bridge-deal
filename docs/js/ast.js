import { Bid, Direction } from './card.js';
class Reject {
    #str;
    #not;
    constructor(str, not) {
        this.#str = str;
        this.#not = not;
    }
    toString() {
        if (this.#not === 1) {
            return `if (${this.#str}) { return false; }`;
        }
        return `if (!(${this.#str})) { return false; }`;
    }
}
class Holder {
    #reject = [];
    #accept = undefined;
    #depth;
    #vars;
    #top;
    constructor(depth, vars, top = false) {
        this.#depth = depth;
        this.#vars = vars;
        this.#top = top;
    }
    add(rule, not) {
        if (typeof rule !== 'string') {
            for (const k of Object.keys(rule)) {
                this.#vars.add(k);
            }
            this.#reject.push(rule);
            return this;
        }
        if (not) {
            this.#reject.push(new Reject(rule, not));
            return this;
        }
        this.#accept = new Accept(rule, this.#vars, this.#depth + 1);
        return this.#accept;
    }
    toString() {
        let ret = '';
        for (const r of this.#reject) {
            if (r instanceof Reject) {
                ret += `${this.indent(1)}${r.toString()}\n`;
            }
            else {
                for (const [k, v] of Object.entries(r)) {
                    if (Array.isArray(v)) {
                        ret += `${this.indent(1)}${k} = [${v.join(', ')}];\n`;
                    }
                    else {
                        ret += `${this.indent(1)}${k} = \`${v}\`;\n`;
                    }
                }
            }
        }
        if (this.#accept) {
            ret += this.#accept.toString();
        }
        else if (!this.#top) {
            ret += `${this.indent(1)}return false;\n`;
        }
        return ret;
    }
    indent(extra = 0) {
        return ''.padEnd((this.#depth + extra) * 2, ' ');
    }
}
class Dir extends Holder {
    #dir;
    constructor(d, vars) {
        super(-1, vars, true);
        this.#dir = d;
    }
    toString() {
        return `dir = deal.${this.#dir};\n${super.toString()}`;
    }
}
class Accept extends Holder {
    #str;
    constructor(str, vars, depth = 0) {
        super(depth, vars);
        this.#str = str;
    }
    toString() {
        let ret = `${this.indent()}if (!(${this.#str})) {\n`;
        ret += super.toString();
        ret += `${this.indent()}}\n`;
        return ret;
    }
}
export class DealRules {
    dealer = Direction.NORTH;
    vuln = undefined;
    bids = [];
    #cur = undefined;
    #dirs = [];
    #vars = new Set(['dir']);
    set dir(d) {
        this.#cur = new Dir(d, this.#vars);
        this.#dirs.push(this.#cur);
    }
    add(rule, not) {
        if (!this.#cur) {
            throw new Error('Set direction first');
        }
        this.#cur = this.#cur.add(rule, not);
    }
    isVar(v) {
        return this.#vars.has(v);
    }
    toString() {
        let ret = '';
        ret += `deal.dealer = '${this.dealer}';\n`;
        if (this.vuln) {
            ret += `deal.vuln = '${this.vuln}'\n`;
        }
        else {
            ret += 'deal.randVuln();\n';
        }
        for (const v of this.#vars) {
            ret += `let ${v} = null;\n`;
        }
        for (const d of this.#dirs) {
            ret += d.toString();
        }
        for (const b of this.bids) {
            ret += `deal.bid(${new Bid(b).serialize()});\n`;
        }
        ret += 'return true;\n';
        return ret;
    }
}
