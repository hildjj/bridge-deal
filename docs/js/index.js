import { Storage } from './storage.js';
import { compressString, decompressString } from './encode.js';
import { Deal } from './card.js';
import { initMonaco } from './monaco.js';
const db = new Storage();
let editor = null;
let model = null;
let monaco = null;
let work = null;
let codeLastStored = NaN;
let state = {
    name: '',
    num: -1n,
    stamp: Date.now(),
};
const history = [];
const add = document.getElementById('btnAdd');
const bidding = document.getElementById('bidding');
const copy = document.getElementById('btnCopy');
const copyFilter = document.getElementById('btnCopyFilter');
const del = document.getElementById('btnDelete');
const diag = document.getElementById('diag');
const diagFilter = document.getElementById('diagFilter');
const diagMessage = document.getElementById('diagMessage');
const error = document.getElementById('error');
const files = document.getElementById('files');
const nxt = document.getElementById('btnNext');
const parResults = document.getElementById('parResults');
const parScore = document.getElementById('parScore');
const prev = document.getElementById('btnPrev');
const rename = document.getElementById('btnRename');
const share = document.getElementById('btnShare');
const tries = document.getElementById('tries');
const holdings = document.querySelectorAll('.holding');
const points = document.querySelectorAll('.points');
const out = document.querySelectorAll('.out');
if (!add || !bidding || !copy || !copyFilter || !del || !diag || !diagFilter ||
    !diagMessage || !error || !files || !nxt || !parResults || !parScore ||
    !prev || !rename || !share || !tries) {
    throw new Error('Element not found');
}
function clear() {
    for (const h of holdings) {
        h.innerText = '';
    }
    for (const p of points) {
        p.innerText = '-';
    }
    for (const o of out) {
        o.innerText = '';
    }
    for (const b of document.querySelectorAll('.bid')) {
        b.remove();
    }
    for (const b of document.querySelectorAll('.bidDir')) {
        b.classList.remove('vuln');
    }
    bidding.style.visibility = 'hidden';
    tries.innerText = '';
    error.innerText = '';
    parResults.innerText = '';
    parScore.innerText = '';
    monaco.editor.removeAllMarkers('web');
}
function snap() {
    const shot = model.createSnapshot();
    let code = '';
    let block = shot.read();
    while (block) {
        code += block;
        block = shot.read();
    }
    return code;
}
function fileOption(text) {
    for (const opt of files.options) {
        if (opt.text === text) {
            return opt;
        }
    }
    return undefined;
}
function fileSelect(text) {
    let i = 0;
    for (const opt of files.options) {
        if (opt.text === text) {
            files.selectedIndex = i;
            break;
        }
        i++;
    }
}
async function nextDeal(num) {
    state.num = num ?? -1n;
    clear();
    if (codeLastStored !== state.stamp) {
        codeLastStored = state.stamp;
        await db.putJS({
            name: state.name,
            stamp: state.stamp,
            code: snap(),
        });
    }
    db.putState(state);
    work?.postMessage(state);
}
function plusMinus(str) {
    if (str.length < 3) {
        return str;
    }
    const firstNum = Number(str.charAt(0));
    const lastNum = Number(str.charAt(str.length - 2));
    return `${str.charAt(0) + str.charAt(str.length - 1)}+${lastNum - firstNum}`;
}
function parContract(str) {
    return str
        .replace(/[0-9]+[CDHSN]/g, n => plusMinus(n))
        .replace(/^(?:NS|EW):/, '');
}
let diagResolve = null;
diag.onclose = () => {
    diagResolve?.(diag.returnValue ? diagFilter.value : null);
};
function diagPrompt(msg, filter) {
    return new Promise(resolve => {
        diag.returnValue = '';
        diagMessage.innerText = msg;
        diagFilter.value = filter;
        diagResolve = resolve;
        diag.showModal();
    });
}
async function gotMessage(e) {
    const typ = e.data.type;
    if (typ === 'ready') {
        const u = new URL(window.location.href);
        let name = u.searchParams.get('name');
        const stamp = u.searchParams.get('stamp');
        const code = u.searchParams.get('code');
        if (name && stamp && code) {
            const ns = parseInt(stamp, 10);
            const newJs = await decompressString(code);
            if (fileOption(name)) {
                const js = await db.getJS(name, ns);
                if (js?.code === newJs) {
                    fileSelect(name);
                    name = null;
                }
                else {
                    name = await diagPrompt('Replace or rename?', name);
                }
            }
            if (name) {
                if (!fileOption(name)) {
                    files.add(new Option(name));
                }
                fileSelect(name);
                state.name = name;
                state.stamp = ns;
                model.setValue(newJs);
            }
        }
        const keys = [...u.searchParams.keys()];
        for (const k of keys) {
            u.searchParams.delete(k);
        }
        if (u.hash) {
            const num = BigInt(`0x${u.hash.replace(/^#/, '')}`);
            window.history.replaceState(num, '', u);
            await nextDeal(num);
        }
        else {
            window.history.replaceState(-1n, '', u);
            await nextDeal();
        }
        return;
    }
    if (typ === 'deal') {
        const deal = Deal.fromJSON(e.data.deal);
        state.num = deal.num;
        for (const h of deal.hands) {
            for (const [suit, cards] of h.suits()) {
                const holding = document.querySelector(`#${h.dir}${suit}`);
                if (holding) {
                    holding.innerText = cards.map(c => c.rank).join('');
                }
            }
            const nm = document.querySelector(`#${h.dir}Name`);
            if (nm) {
                nm.innerText = h.name;
            }
            const pts = document.querySelector(`#${h.dir}Points`);
            if (pts) {
                pts.innerText = String(h.points);
            }
        }
        const vulnDirs = {
            All: ['north', 'south', 'east', 'west'],
            None: [],
            NS: ['north', 'south'],
            EW: ['east', 'west'],
        }[deal.vuln] ?? [];
        for (const v of vulnDirs) {
            const dirDiv = document.querySelector(`#${v}Bid`);
            if (dirDiv) {
                dirDiv.classList.add('vuln');
            }
        }
        if (deal.bids.length > 0) {
            let i = 0;
            for (const bid of deal.bids) {
                const div = document.createElement('div');
                div.innerText = bid.toString(true);
                if (!i++) {
                    div.style.gridColumnStart = {
                        west: '1',
                        north: '2',
                        east: '3',
                        south: '4',
                    }[deal.dealer];
                }
                div.classList.add('bid');
                if (bid.level !== 0) {
                    div.classList.add('notPass');
                }
                if (bid.alert) {
                    div.classList.add('alert');
                }
                if (bid.description) {
                    div.title = bid.description;
                }
                bidding.appendChild(div);
            }
            bidding.style.visibility = 'visible';
        }
        const u = new URL(window.location.href);
        u.hash = state.num.toString(16);
        window.history.replaceState(state.num, '', u);
        document.title = `Bridge Dealer - ${u.hash}`;
        copy.classList.remove('md-inactive');
        copy.disabled = false;
        if (e.data.tries) {
            tries.innerHTML = `Found deal after <code>${e.data.tries.toLocaleString()}</code> tries.`;
        }
        return;
    }
    if (typ === 'tricks') {
        if (e.data.sess.sockref !== state.num.toString(16)) {
            return;
        }
        let i = 0;
        for (const dir of ['North', 'South', 'East', 'West']) {
            for (const trump of ['NT', 'Spades', 'Hearts', 'Diamonds', 'Clubs']) {
                const tricks = parseInt(e.data.sess.ddtricks[i++], 16);
                const outDir = document.querySelector(`#out${dir}${trump}`);
                if (outDir) {
                    outDir.innerText = (tricks > 6) ? String(tricks - 6) : '-';
                }
            }
        }
        parResults.innerText = parContract(e.data.contractsNS);
        parScore.innerText = e.data.scoreNS;
        return;
    }
    if (typ === 'error') {
        if (e.data.location) {
            error.innerText = 'Deal syntax error';
            const { start, end } = e.data.location;
            monaco.editor.setModelMarkers(model, 'web', [{
                    startLineNumber: start.line,
                    startColumn: start.column,
                    endLineNumber: end.line,
                    endColumn: end.column,
                    message: e.data.error.message,
                    severity: monaco.MarkerSeverity.Error,
                }]);
        }
        else {
            error.innerText = e.data.error.message;
        }
        return;
    }
    error.innerText = `Invalid message type: "${typ}"`;
}
db.init().then(async () => {
    ([editor, model, monaco] = await initMonaco('monaco', () => {
        state.stamp = Date.now();
        monaco.editor.removeAllMarkers('web');
    }));
    editor.addAction({
        id: 'bridge-deal-next',
        label: 'Next Deal',
        keybindings: [
            monaco.KeyMod.WinCtrl | monaco.KeyCode.KeyN,
        ],
        contextMenuGroupId: 'navigation',
        run: () => nextDeal(),
    });
    state = await db.getState();
    const names = await db.getJSnames();
    names.sort();
    let i = 0;
    let index = 0;
    for (const n of names) {
        const selected = (n === state.name);
        files.add(new Option(n));
        if (selected) {
            index = i;
        }
        i++;
    }
    files.selectedIndex = index;
    const js = await db.getJS(state.name, state.stamp);
    if (js) {
        model.setValue(js.code);
    }
    work = new Worker('./js/worker.js');
    work.onmessage = gotMessage;
});
prev.onclick = () => {
    const last = history.pop();
    if (last !== undefined) {
        if (history.length === 0) {
            prev.classList.add('md-inactive');
            prev.disabled = true;
        }
        nextDeal(last);
    }
};
copy.onclick = () => {
    navigator.clipboard.writeText(window.location.href);
};
nxt.onclick = () => {
    prev.classList.remove('md-inactive');
    prev.disabled = false;
    history.push(state.num);
    nextDeal();
};
rename.onclick = async () => {
    const name = await diagPrompt('New name?', state.name);
    if (name !== null) {
        const opt = files.item(files.selectedIndex);
        if (opt) {
            opt.text = name;
        }
        await db.delJS(state.name);
        state.name = name;
        state.stamp = Date.now();
        nextDeal();
    }
};
copyFilter.onclick = async () => {
    const name = await diagPrompt('Save as?', state.name);
    if (!name) {
        return;
    }
    state.name = name;
    state.stamp = Date.now();
    files.add(new Option(name));
    files.selectedIndex = files.options.length - 1;
    nextDeal();
};
share.onclick = async () => {
    const u = new URL(window.location.href);
    u.hash = '';
    u.searchParams.append('name', state.name);
    u.searchParams.append('stamp', String(state.stamp));
    u.searchParams.append('code', await compressString(snap()));
    await navigator.clipboard.writeText(u.href);
};
async function newFilter() {
    let defName = 'Filter';
    for (let i = 0; i < 1000; i++) {
        defName = `Filter ${i}`;
        if (!fileOption(defName)) {
            break;
        }
    }
    const name = await diagPrompt('New name?', defName) || defName;
    files.add(new Option(name));
    model.setValue('return true');
    files.selectedIndex = files.options.length - 1;
    state.name = name;
    state.stamp = Date.now();
    await nextDeal();
}
del.onclick = async () => {
    let opt = files.item(files.selectedIndex);
    if (opt) {
        const old = opt.text;
        files.remove(files.selectedIndex);
        files.selectedIndex = 0;
        await db.delJS(old);
        if (files.options.length > 0) {
            opt = files.item(0);
            if (opt) {
                const code = await db.getJS(opt.text, Date.now());
                if (!code) {
                    error.innerText = `Unknown filter "${opt.text}"`;
                    return;
                }
                state.stamp = code.stamp;
                state.name = code.name;
                model.setValue(code.code);
                nextDeal();
            }
        }
        else {
            await newFilter();
        }
    }
};
add.onclick = newFilter;
window.onkeydown = (ev) => {
    if (ev.isComposing || ev.keyCode === 229) {
        return true;
    }
    if (ev.ctrlKey && !ev.altKey && !ev.metaKey) {
        switch (ev.key) {
            case 'n':
                nxt.click();
                break;
            case 'p':
                prev.click();
                break;
            default:
                return true;
        }
        return false;
    }
    return true;
};
files.onchange = async (ev) => {
    await db.putJS({
        name: state.name,
        stamp: state.stamp,
        code: snap(),
    });
    const name = files.item(files.selectedIndex)?.value;
    if (name) {
        const code = await db.getJS(name, Date.now());
        if (code) {
            state.name = name;
            state.stamp = code.stamp;
            model?.setValue(code.code);
            editor?.setScrollTop(0, 1);
            nextDeal();
        }
    }
};
