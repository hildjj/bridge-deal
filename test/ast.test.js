import {Direction, Vuln} from '../docs/js/card.js';
import {default as assert, equal, throws} from 'node:assert';
import {DealRules} from '../docs/js/ast.js';
import snap from 'snappy-snaps';
import {test} from 'node:test';

test('DealRules', async() => {
  const d = new DealRules();
  assert(d);
  throws(() => d.add('foo', 0));
  d.dealer = Direction.SOUTH;
  d.vuln = Vuln.ALL;
  d.dir = Direction.SOUTH;
  d.add({foo: 1}, 0);
  d.add({points: [4, 3, 2, 1]}, 0);
  equal(d.isVar('foo'), true);
  equal(d.isVar('bar'), false);
  d.add('dir.balancedNoM()', 1);
  d.add('dir.points === 17', 0);
  // eslint-disable-next-line no-template-curly-in-string
  d.bids.push('P: ${foo}');
  const code = d.toString();
  const expected = await snap('DealRules', code);
  equal(code, expected);
  // eslint-disable-next-line no-template-curly-in-string
  d.bids.push('P: ${bar}');
  throws(() => d.toString());
  d.bids.pop();
});
