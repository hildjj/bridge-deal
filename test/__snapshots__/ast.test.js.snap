// Data Snap v1

exports[`DealRules`] = {
  "data": "deal.dealer = 'south';\ndeal.vuln = 'All'\nlet dir = null;\nlet foo = null;\nlet points = null;\ndir = deal.south;\nfoo = `1`;\npoints = [4, 3, 2, 1];\nif (dir.balancedNoM()) { return false; }\nif (!(dir.points === 17)) {\n  return false;\n}\ndeal.bid({level: 0,description: `${foo}`,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}
